extension/src/popup/helpers/account.ts

```diff
@@ -15,3 +24,105 @@ export const sortBalances = (balances: Balances) => {
 
   return collection;
 };
+
+export const getIsPayment = (type: Horizon.OperationResponseType) =>
+  [
+    Horizon.OperationResponseType.payment,
+    Horizon.OperationResponseType.pathPayment,
+    Horizon.OperationResponseType.pathPaymentStrictSend,
+  ].includes(type);
+
+export const getIsSwap = (operation: HorizonOperation) =>
+  operation.type_i === 13 && operation.source_account === operation.to;
+
+interface SortOperationsByAsset {
+  operations: Array<HorizonOperation>;
+  balances: Array<Types.AssetBalance | Types.NativeBalance>;
+}
+
+export interface AssetOperations {
+  [key: string]: Array<HorizonOperation>;
+}
+
+export const sortOperationsByAsset = ({
+  balances,
+  operations,
+}: SortOperationsByAsset) => {
+  const assetOperationMap = {} as AssetOperations;
+
+  balances.forEach((bal) => {
+    if (bal.token) {
+      const issuer = "issuer" in bal.token ? bal.token.issuer.key : "";
+      assetOperationMap[getCanonicalFromAsset(bal.token.code, issuer)] = [];
+    }
+  });
+
+  operations.forEach((op) => {
+    if (getIsPayment(op.type)) {
+      Object.keys(assetOperationMap).forEach((assetKey) => {
+        const asset = getAssetFromCanonical(assetKey);
+        const assetCode = asset.code === "XLM" ? "native" : asset.code;
+        const assetIssuer = asset.issuer;
+
+        if (
+          (op.asset_code === assetCode && op.asset_issuer === assetIssuer) ||
+          op.asset_type === assetCode
```

Comment:
this case would only be for XLM right? can we hardcode that here?
```suggestion
          op.asset_type === "native"
```

Reply:
Ah this is clearer, but unfortunately does not work here. Since it's essentially a nested for loop, we need to make sure that not only does `asset_type === "native"` but also that the `asset` we are looking at it is "native"

In the truthy scenario, the operation loop's `asset_type` is the same as the asset loop's `assetCode`